ci: remove 3rd party js from windows dll gha job

m3dwards opened the pull request:
The windows job uses the external dependency `ilammy/msvc-dev-cmd` which runs javascript. We use this to put various tools on the path such as `MSBuild.exe` and `mt.exe`. We can remove this dependency and use `vswhere.exe` directly to find these tools and create a "[Developer command prompt](https://github.com/microsoft/vswhere/wiki/Start-Developer-Command-Prompt#using-powershell)" as someone would on their dev machine.

While in this area of the code, this PR also runs some additional manifest checks on the windows binaries.

Fixes: #32508

hebasto commented:
Concept ACK.

Thank you for taking this!

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -185,11 +185,17 @@ jobs:
+          $vswherePath = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe"
+          $msbuildPath = & "$vswherePath" -latest -requires Microsoft.Component.MSBuild -find MSBuild\**\Bin\MSBuild.exe
+          if (-not $msbuildPath) {
+            throw "MSBuild not found"
```
Why use an exception? Wouldn’t exiting with an error code suffice?

m3dwards replied:
I spose I'd ask the question the other way round, why not the exception? It would be an exceptional circumstance if MSBuild could not be found.

m3dwards replied:
No, this [doc](https://docs.github.com/en/actions/sharing-automations/creating-actions/setting-exit-codes-for-actions) just says use any exit code other than 0 (which throw does do anyway).

I've tested throw works here: https://github.com/m3dwards/bitcoin/actions/runs/15050358923/job/42303402427

All that said, I'm really not wedded to it and happy to change to a simple `exit 1` if you prefer.

hebasto replied:
I have no preferences either.

hebasto reviewed (approved):
ACK 2549a1075521202ee0d9fe069ba1062563ee6387, tested locally on Windows 11. Additionally tested in my personal repo with invalidated vcpkg binary cache.

m3dwards commented on .github/workflows/ci.yml:
```diff
@@ -185,11 +185,17 @@ jobs:
+          }
+          $msbuildDir = Split-Path $msbuildPath -Parent
+          Write-Host "MSBuild found at $msbuildPath"
+          "$msbuildDir" | Out-File -FilePath $env:GITHUB_PATH -Encoding utf8 -Append
```
Will try without.

m3dwards replied:
Worked, force pushed [449cb9e](https://github.com/bitcoin/bitcoin/commit/449cb9e8985b350a96ddad74a9997ac95118edd0)

davidgumberg commented:
It might be better to use a more general solution for setting up VS prompts, this would also work once #32396 is merged.

```yml
      - name: Set up VS Developer Prompt
        shell: pwsh
        run: |
          $vswherePath = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe"
          $installationPath = & 'C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe' -latest -property installationPath
          if ($installationPath -and (test-path "$installationPath\Common7\Tools\vsdevcmd.bat")) {
            & "${env:COMSPEC}" /s /c "`"$installationPath\Common7\Tools\vsdevcmd.bat`" -no_logo && set" | foreach-object {
              $name, $value = $_ -split '=', 2
              echo "$name=$value" >> $env:GITHUB_ENV
            }
          }
```
(https://github.com/davidgumberg/bitcoin/commit/83c53bfb58bf5451dccee753ace1c7464ea4ac5f)


This snippet was mostly taken from this `vswhere` wiki article: https://github.com/microsoft/vswhere/wiki/Start-Developer-Command-Prompt#using-powershell

hebasto commented:
> It might be better to use a more general solution for setting up VS prompts, this would also work once #32396 is merged.

@m3dwards 

Could you please rebase this PR on top of https://github.com/bitcoin/bitcoin/pull/32396?

m3dwards commented:
Rebased and taken @davidgumberg's great suggestion.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -185,11 +185,17 @@ jobs:
+          $installationPath = & 'C:\Program Files (x86)\Microsoft Visual Studio\Installer\vswhere.exe' -latest -property installationPath
+          if ($installationPath -and (test-path "$installationPath\Common7\Tools\vsdevcmd.bat")) {
+            & "${env:COMSPEC}" /s /c "`"$installationPath\Common7\Tools\vsdevcmd.bat`" -no_logo && set" | foreach-object {
+              $name, $value = $_ -split '=', 2
+              echo "$name=$value" >> $env:GITHUB_ENV
```
This part seems too invasive to me. Please compare the list of environment variables in the "Get tool information" step on the master branch with those in this PR.

hebasto replied:
> The only tool we need is msbuild.exe...

We also use now `mt.exe`: https://github.com/bitcoin/bitcoin/blob/dc403f0c13e1ceee3d4651b4056ff2bad47d0315/.github/workflows/ci.yml#L246-L253

hebasto replied:
> Shall I just put it back?

Perhaps updating only the `Path` variable will be sufficient?

hebasto replied:
> I will look at using the top commit hash for cache invalidation. If I've understood you correct the current cache key is:
> 
> `hashFiles('cmake_version', 'msbuild_version', 'toolset_version', 'vcpkg.json')`
> 
> we think this will be sufficient:
> 
> `hashFiles('vcpkg_commit_hash', 'vcpkg.json')`

Yes, I thin so.

> Which then means we can drop the vswhere tool finding.

Right. However, we still need to locate `mt.exe`.

m3dwards replied:
@hebasto compelling enough to ACK as is?

davidgumberg commented:
This approach here should also be applied to replace the `Find mt.exe tool` step in the `Windows, test cross-built` job.

hebasto reviewed (approved):
ACK ff2e35f6315cc6b912f68fe85103575d751bd2bc.

While touching this code, it seems reasonable to also address the other @davidgumberg's [comment](https://github.com/bitcoin/bitcoin/pull/32634#issuecomment-2918081325):
> Might be nice to have a CI check like [davidgumberg@1b98160](https://github.com/davidgumberg/bitcoin/commit/1b9816052b1d8ad1d9c17945530e14ead79fce33) to prevent future regressions, but OK for a separate PR if out of scope.

m3dwards commented:
> While touching this code, it seems reasonable to also address the other @davidgumberg's [comment](https://github.com/bitcoin/bitcoin/pull/32634#issuecomment-2918081325):

I don't mind adding that.

m3dwards commented on .github/workflows/ci.yml:
```diff
@@ -356,19 +362,24 @@ jobs:
-          "MT_EXE=${sdk_dir}bin\${sdk_latest}\x64\mt.exe" >> $env:GITHUB_ENV
+          $vswherePath = "${env:ProgramFiles(x86)}\Microsoft Visual Studio\Installer\vswhere.exe"
+          $installationPath = & $vswherePath -latest -property installationPath
+          if ($installationPath -and (test-path "$installationPath\Common7\Tools\vsdevcmd.bat")) {
```
Don't worry, I'm reworking it to do more manifest checks anyway so changes at this point are fine.

Will look at this.

m3dwards commented:
> While touching this code, it seems reasonable to also address the other @davidgumberg's [comment](https://github.com/bitcoin/bitcoin/pull/32634#issuecomment-2918081325):

Added

Dakota48423 commented:
Leveta kalerei?
DakotaAllen

On Mon, Jun 2, 2025, 5:23 AM Max Edwards ***@***.***> wrote:

> *m3dwards* left a comment (bitcoin/bitcoin#32513)
> <https://github.com/bitcoin/bitcoin/pull/32513#issuecomment-2929886599>
>
> While touching this code, it seems reasonable to also address the other
> @davidgumberg <https://github.com/davidgumberg>'s comment
> <https://github.com/bitcoin/bitcoin/pull/32634#issuecomment-2918081325>:
>
> Added
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/32513#issuecomment-2929886599>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BMTNS5KQXHV3MGOHD64VP4D3BQQ23AVCNFSM6AAAAAB5GE6TUKVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMRZHA4DMNJZHE>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

davidgumberg commented on .github/workflows/ci.yml:
```diff
@@ -370,12 +386,28 @@ jobs:
+
+            Write-Host "Checking $exeName"
+            & mt.exe -nologo -inputresource:$_.FullName -validate_manifest
+            if ($LASTEXITCODE -ne 0) {
```
If I'm not mistaken, this `if` branch won't be reached if the line above fails.

m3dwards replied:
Default behaviour in powershell is to continue on error (unless it's a terminating error like throw). When calling a powershell cmdlet you can set `$ErrorActionPreference = 'Stop'` but being as this is a call to an external process I don't believe that approach works here, hence why checking the exit code.

m3dwards replied:
Doesn't appear to be turned off on our CI and I think it should be.

In Powershell 7.4 and up there is now a setting `PSNativeCommandUseErrorActionPreference` which should make calling native commands respect the `ErrorActionPreference` so we should set both of those, globally if possible. I will look into it.

m3dwards replied:
I'm wrong about powershell cmdlets: https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#exit-codes-and-error-action-preference

Github sets `ErrorActionPreference` to `stop` but it doesn't appear that they set `PSNativeCommandUseErrorActionPreference` which is why calling native executables (like mt.exe) can still fail silently.

hebasto replied:
> > Default behaviour in powershell is to continue on error
> 
> Do we have that turned off for all of our CIs?

We did: https://github.com/bitcoin/bitcoin/blob/e872a566f251c73908de8b6d243c94a6679c2eac/.github/workflows/ci.yml#L24-L28

m3dwards replied:
@hebasto but there are a few places where we use pwsh, which will still fail fast if using exclusively cmdlets but doesn't appear to fail when calling native executables.

Here is a commit to demonstrate it won't always failfast: https://github.com/bitcoin/bitcoin/commit/4bf10310fb4ec4d6c2f479f008c487cca1f90c27

and the job run that didn't fail: https://github.com/m3dwards/bitcoin/actions/runs/15415032095/job/43375709475

Suggest we set `PSNativeCommandUseErrorActionPreference` to `true` everywhere we use pwsh.

m3dwards replied:
@davidgumberg force pushed a change so that it behaves how you expected it to: https://github.com/bitcoin/bitcoin/compare/a453e66c827f260da597bf70273bcd0da2313774..7f8393ad745e6bc44dc9481428ea5979ed31a3c9

davidgumberg commented on .github/workflows/ci.yml:
```diff
@@ -211,11 +211,15 @@ jobs:
+          }
 
       - name: Get tool information
         shell: pwsh
```
```suggestion
        shell: pwsh -Command "$PSVersionTable; $PSNativeCommandUseErrorActionPreference = $true; $ErrorActionPreference = 'Stop'; & '{0}'"
```

davidgumberg reviewed (commented):
Nice! Looks good, there are still two places where `shell: pwsh` is used without your ` -Command "$PSVersionTable; $PSNativeCommandUseErrorActionPreference = $true; $ErrorActionPreference = 'Stop'; & '{0}'"` suffix:


https://github.com/bitcoin/bitcoin/blob/c44fb4facfd3a5db8aa1bd2c7f5506a1477b0d2b/.github/workflows/ci.yml#L224-L225

and

Not a big deal for this since it's only using powershell cmdlets now, but for the sake of consistency
https://github.com/bitcoin/bitcoin/blob/c44fb4facfd3a5db8aa1bd2c7f5506a1477b0d2b/.github/workflows/ci.yml#L431-L432

m3dwards committed:
ci: remove 3rd party js from windows dll gha job

We can use vswhere.exe directly to create a vs developer
prompt and so can remove this third party dependency.

Co-authored-by: David Gumberg <davidzgumberg@gmail.com>

m3dwards commented:
> On the [master](https://github.com/bitcoin/bitcoin/actions/runs/17921155333/job/50956194860) branch:
> 
> ```
> VSCMD_ARG_HOST_ARCH: x64
> VSCMD_ARG_TGT_ARCH: x64
> ```
> 
> However, with this [PR](https://github.com/bitcoin/bitcoin/actions/runs/17807365691/job/50622408408):
> 
> ```
> VSCMD_ARG_HOST_ARCH: x86
> VSCMD_ARG_TGT_ARCH: x86
> ```

What a great catch! Sorted it in this [run](https://github.com/m3dwards/bitcoin/actions/runs/18015235679/job/51258366346), will update PR shortly

m3dwards committed:
ci: use a more generic way of finding mt.exe

This sets up a vs developer command prompt and should hopefully should
be more resilient to upstream changes

Co-authored-by: David Gumberg <davidzgumberg@gmail.com>

m3dwards committed:
ci: Check windows manifests for all executables

The other executables have manifests and these should be checked in
addition to bitcoind. Skipping fuzz.exe, bench_bitcoin.exe and
test_bitcoin-qt.exe as they do not have manifests.

davidgumberg reviewed (approved):
crACK 156927903d64297

Nice! Was bad that we were depending on 3rd party JS before this.

This branch also extends the executable manifest check to all binaries, which would have caught the missing manifest in  #31375 which #32634 caught.

The CI failure seems spurious and unrelated.

m3dwards commented:
Thanks @davidgumberg!

@hebasto should have addressed your comments.

hebasto reviewed (approved):
ACK 156927903d64297500dd73380908c654b07bfb1a.
